[RA1 Ch03] 3.3.3 Cloud partitioning: Host aggregates, Availability Zones (#1270)
cntt-n/CNTT: doc/ref_arch/openstack/chapters/chapter03.md

OpenStack offers multi-tenancy by means of resource (compute, network and storage)separation via projects. OpenStack offers ways to share virtual resources between projects while maintaining logical separation. As an example, traffic separation is provided by creating different VLAN ids for neutron networks of different projects. As another example, if host separation is needed, nova scheduler offers AggregateMultiTenancyIsolation scheduler filter to separate projects in host aggregates. Thus, if a host in an aggregate is configured for a particular project, only the instances from that project are placed on the host. Overall, tenant isolation ensures that the resources of a project are not affected by resources of another project.

<a name="3.3.3"></a>
### 3.3.3. Cloud partitioning: Host Aggregates, Availability Zones

> Cloud administrators can partitioned the hosts within an OpenStack cloud using 2 mechanisms: Host Aggregates and Availability Zones.

Reviewer: @karinesevilla As you know AZs are host aggregates that are made visible to tenants and with the restriction that hosts can only belong to one AZ. (Applies also to later content). So, actually there is only onemechanism for partitioning.

Typo: "Cloud administrators can partitioned ..." ---> "Cloud administrators can partition ..."

Author: @goyal You are right, I modified the text accordingly with a second commit

> A Host Aggregate is a group of hosts (compute nodes) with specific characteristics, software and/or hardware properties. Example would be a Host Aggregate created for specific hardware or performances. The administrator assign key-value pairs to groups of compute nodes, this information will be used to schedule VMs. A host can belong to multiple host aggregates. Host Aggregates are not explicitly exposed to tenants.

Reviewer: "The administrator assign key-value pairs to groups of compute nodes, this information will be used to schedule VMs." -- I think the key-value properties are assigned to the host aggregates.

Typo: ""The administrator assign ..." ---> "The administrator assigns ..."

Reviewer 2: @pgoyal01 , Can we mention "with specific characteristics with the same specification"

Reviewer 2: @pgoyal01 , we have a deployment, one HA for DPDK workload, 2nd HA for ovs deployment ,...

Author: @goyal I'm not sure to catch your point. For me "The administrator assign key-value pairs to groups of compute nodes" is equivalent to "key-value properties are assigned to the host aggregates".
Typo fixed

@ASawwaf Yes, I can add with the same specification
-> done in commit 3

Reviewer: @karinesevilla I leave it up to you to decide whether to use "groups of compute nodes" or "host aggregate". Thanks

Author: @pgoyal01 Got it. I change the sentence if it's clearer.

> A Host Aggregate is a group of hosts (compute nodes) with specific characteristics and with the same specifications, software and/or hardware properties. Example would be a Host Aggregate created for specific hardware or performances. The administrator assigns key-value pairs to groups of compute nodes, this information will be used to schedule VMs. A host can belong to multiple host aggregates. Host Aggregates are not explicitly exposed to tenants.

Reviewer 3: Typo 'performances' should be 'performance characteristics'

Author: Done

> A Host Aggregate is a group of hosts (compute nodes) with specific characteristics and with the same specifications, software and/or hardware properties. Example would be a Host Aggregate created for specific hardware or performance characteristics. The administrator assigns key-value pairs to groups of compute nodes, this information will be used to schedule VMs. A host can belong to multiple host aggregates. Host Aggregates are not explicitly exposed to tenants.

Reviewer 4: Small grammar nit:
```suggestion
A Host Aggregate is a group of hosts (compute nodes) with specific characteristics and with the same specifications, software and/or hardware properties. Example would be a Host Aggregate created for specific hardware or performance characteristics. The administrator assigns key-value pairs to groups of compute nodes that is then used when scheduling VMs. A host can belong to multiple host aggregates. Host Aggregates are not explicitly exposed to tenants.
```

> Availability Zones (AZs) are another mechanism for partitioning. They are exposed to tenants. They are defined by attaching specific metadata information to an aggregate, making the aggregate visible for tenants. Hosts can only be in a single Availability Zone. By default a host is part of a default Availability Zone, even if it doesn’t belong to an aggregate. Availability Zones are used to provide resiliency and fault tolerance for VM deployments, by means of physical hosting distribution of Compute Nodes in separate racks with separate power supply and eventually in different rooms. They permit rolling upgrades – an AZ at a time upgrade with enough time between AZ upgrades to allow recovery of tenant workloads on the upgraded AZ. Finally, AZs can accommodate hosts with special hardware and software characteristics, for example, hosts with hardware accelerators.

Reviewer: "Availability Zones are used to provide resiliency and fault tolerance for VM deployments, by means of physical hosting distribution of Compute Nodes in separate racks with separate power supply and eventually in different rooms." Some of the characteristics are operator dependent and not necessarily requirements for AZ.

Also, the resiliency is for the workloads and not a VM.

Reviewer 2: @pgoyal01 , thanks for extensive definition, I like to go with workload instead of VM in " provide resiliency and fault tolerance for VM deployments"

Author: @pgoyal01 
Some of the characteristics are operator dependent and not necessarily requirements for AZ -> Agree, I will mention it as an example
@goyal @ASawwaf Agree, VM changed by workload
Done -> Commit 4

> Availability Zones (AZs) rely on Host Aggregates and make the partiotioning visible to tenants. They are defined by attaching specific metadata information to an aggregate, making the aggregate visible for tenants. Hosts can only be in a single Availability Zone. By default a host is part of a default Availability Zone, even if it doesn’t belong to an aggregate. Availability Zones are used to provide resiliency and fault tolerance for VM deployments, by means of physical hosting distribution of Compute Nodes in separate racks with separate power supply and eventually in different rooms. They permit rolling upgrades – an AZ at a time upgrade with enough time between AZ upgrades to allow recovery of tenant workloads on the upgraded AZ. Finally, AZs can accommodate hosts with special hardware and software characteristics, for example, hosts with hardware accelerators.

Reviewer 3: Typo 'partiotioning"

Author: @iangardner22 Thanks!
done in commit 4

> Availability Zones (AZs) rely on Host Aggregates and make the partitioning visible to tenants. They are defined by attaching specific metadata information to an aggregate, making the aggregate visible for tenants. Hosts can only be in a single Availability Zone. By default a host is part of a default Availability Zone, even if it doesn’t belong to an aggregate. Availability Zones can be used to provide resiliency and fault tolerance for workloads deployments, for example by means of physical hosting distribution of Compute Nodes in separate racks with separate power supply and eventually in different rooms. They permit rolling upgrades – an AZ at a time upgrade with enough time between AZ upgrades to allow recovery of tenant workloads on the upgraded AZ. Finally, AZs can accommodate hosts with special hardware and software characteristics, for example, hosts with hardware accelerators.

Reviewer 4: When would one use AZs for special hardware and software characteristics, compared to the host aggregate and metadata approach above?

Reviewer: @kedmison W.r.t. "When would one use AZs for special hardware and software characteristics, compared to the host aggregate and metadata approach above?" when we want to expose to tenants; as you know tenant do not know anything about host-aggregates.

Host aggregates are grouping of resources that have some common characteristic; the resource can belong to multiple host aggregates. Host aggregates do not partition resources in the same fashion as an AZ.

We can create host aggregates that correspond to instance profiles (viz., flavor geometry and capabilities). But we can create an AZ with hosts configured to support many instance profiles.

W.r.t. AZs able to accommodate hosts with special hardware and software (while not necessary) but can be done where we need to earmark these special resources, potentially expensive, to certain tenants.

Reviewer 4: This sentence "Finally, AZs can accommodate hosts with special hardware and software characteristics, for example, hosts with hardware accelerators." seems to be very confusing since that is pretty much what host aggregates and NFVI profiles/Openstack flavours are for: identifying special hosts and scheduling work on them.

While it may be possible to use AZs for this, perhaps this pattern shouldn't appear in the RA document, as CNTT is largely organizing around exposing hardware and capabilities through NFVI profiles.
```suggestion
Availability Zones (AZs) rely on Host Aggregates and make the partitioning visible to tenants. They are defined by attaching specific metadata information to an aggregate, making the aggregate visible for tenants. Hosts can only be in a single Availability Zone. By default a host is part of a default Availability Zone, even if it doesn’t belong to an aggregate. Availability Zones can be used to provide resiliency and fault tolerance for workloads deployments, for example by means of physical hosting distribution of Compute Nodes in separate racks with separate power supply and eventually in different rooms. They permit rolling upgrades – an AZ at a time upgrade with enough time between AZ upgrades to allow recovery of tenant workloads on the upgraded AZ. 
```

Reviewer: @kedmison That is fine. Thanks

@karinesevilla Up to you to incorporate the language. Thanks.

> An over use of Host Aggregates and Availability Zones can result in a granular partition the cloud and, hence, operational complexities and inefficiencies.

Reviewer: do you want to add some examples of "inefficiencies" such as in resource utilisation, operational complexities and inefficiencies?

Author: @pgoyal01 Yes, it will be good
